File: go/0092_reverse_linked_list_ii.go
func reverseBetween(head *ListNode, m int, n int) *ListNode {
	if head == nil || head.Next == nil || m == n {
		return head
	}

	var left, right, start, prev, cur, next *ListNode
	cur = head
	next = cur.Next

	i := 1
	for { // iterate until m
		if i == m {
			break
		}
		i++
		prev = cur
		cur = next
		next = next.Next
	}

	left, start = prev, cur
	for { // reverse m to n
		if i > n {
			break
		}
		i++
		cur.Next = prev
		prev = cur
		cur = next
		if next != nil {
			next = next.Next
		}
	}
	right = cur

	// connect left, right and adjust head
	if left == nil {
		head = prev
	} else {
		left.Next = prev
	}
	start.Next = right
	return head
}

File: python/0092_reverse_linked_list_II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """

        i = 1
        start = head
        prevStart = None
        while i < m:
            prevStart = start
            start = start.next
            i += 1

        prev = None
        cur = start
        next = cur.next
        while i <= n:
            cur.next = prev
            prev = cur
            cur = next
            if cur:
                next = cur.next
            i += 1

        start.next = cur
        if prevStart:
            prevStart.next = prev
            return head
        return prev





